crates/worker/src/purgatory.rs

```diff
@@ -64,57 +91,83 @@ impl Drop for Inner {
 }
 
 impl Inner {
-    fn new(handle: NonNull<JoinHandle<()>>, state: State) -> Self {
+    fn new(
+        handle: NonNull<JoinHandle<()>>,
+        state: State,
+        db_client: DynamoDBClient,
+        s3_client: S3Client,
+    ) -> Self {
         Self {
-            handle,
             state,
+            s3_client,
+            db_client,
+            handle,
             _marker: PhantomData,
         }
     }
 
-    pub fn purge(&mut self) {
-        todo!()
+    fn add_record(&mut self, id: Uuid, result: TaskResult) {
+        pub const DURATION_TO_PURGE: u64 = 60 * 5; // 5 minutes
+        let to_purge_timestampe = timestamp() + DURATION_TO_PURGE;
+
+        self.state.task_status.insert(id, Status::Done(result));
+        self.state
+            .expiration_timestamps
+            .push((id, to_purge_timestampe));
     }
 
-    // TODO: replace with Self::purge
-    // pub async fn supervisor(
-    //     db_client: DynamoDBClient,
-    //     expiration_timestamps: Arc<Mutex<Vec<(Uuid, Timestamp)>>>,
-    // ) {
-    //     loop {
-    //         let now = timestamp();
-    //
-    //         let to_delete = {
-    //             let mut to_delete = vec![];
-    //             let mut expiration_timestamps = expiration_timestamps.lock().await;
-    //             expiration_timestamps.retain(|&(uuid, expiration)| {
-    //                 if expiration < now {
-    //                     to_delete.push(uuid);
-    //                     false
-    //                 } else {
-    //                     true
-    //                 }
-    //             });
-    //
-    //             to_delete
-    //         };
-    //
-    //         for uuid in to_delete {
-    //             db_client.delete_item(uuid.to_string()).await;
-    //         }
-    //
-    //         sleep(Duration::from_millis(2000)).await;
-    //     }
-    // }
-}
+    pub async fn purge(&mut self) {
```

Comment:
I think it's better to make purge return error, and then handle it in the root, it's just to avoid a lot of unwraps and control it more

Reply:
It will. I reckoned that it is just out of scope of this pr. This is related to retriable clients #182 as mentioned in left TODOs